src/supervisor/mod.rs

```diff
@@ -64,16 +71,39 @@ pub struct Spec {
     pub meta: SpecMeta,
 }
 
+/// A value that gets returned when running a Supervisor record
+enum SupervisorResult {
```

Comment:
This is way better :+1: still complex, but as you said this is inherent complexity. And now it has nice names!